repo/blob/s3/s3_storage.go

```diff
@@ -160,10 +165,15 @@ func (s *s3Storage) putBlob(ctx context.Context, b blob.ID, data blob.Bytes) (ve
 
 	if errors.Is(err, io.EOF) && uploadInfo.Size == 0 {
 		// special case empty stream
-		_, err = s.cli.PutObject(ctx, s.BucketName, s.getObjectNameString(b), bytes.NewBuffer(nil), 0, minio.PutObjectOptions{
-			ContentType:  "application/x-kopia",
-			StorageClass: storageClass,
-		})
+		putOpts = minio.PutObjectOptions{
+			ContentType: "application/x-kopia",
+		}
+		if blobOpts != nil {
+			putOpts.StorageClass = blobOpts.StorageClass
+			putOpts.RetainUntilDate = blobOpts.RetainUntilDate
```

Comment:
Specifying a `RetainUntilDate` may fail when:
- The bucket is not locked, that is, object locking is not enabled.
- The client credentials do not have permissions to specify a `RetainUntilDate` (actually, I'm not sure if there's a specific permission for this. There is definitely a permission for being able to extend the `RetainUntilDate`)

In the case of failure, the operation should be retried without a `RetainUntilDate` set. Similar to how `putOpts.SendContentMd5` is handled above.